### PR TITLE
GSW-1838 refactor: tick

### PR DESCRIPTION
**Pull request by r3v4s**
## refactor
1. pool's tick.gno
> test case already exists in current main

### CODE REVIEW

File: pool/tick.gno [hunk 1]
```diff
@@ -16,41 +16,23 @@ func tickTickSpacingToMaxLiquidityPerTick(tickSpacing int32) *u256.Uint {
 	maxTick := (consts.MAX_TICK / tickSpacing) * tickSpacing
 	numTicks := uint64((maxTick-minTick)/tickSpacing) + 1
 
-	_maxUint128 := u256.MustFromDecimal(consts.MAX_UINT128)
-	return new(u256.Uint).Div(_maxUint128, u256.NewUint(numTicks))
+	return new(u256.Uint).Div(consts.MAX_UINT128.Clone(), u256.NewUint(numTicks))
```
Comment (dongwon8247) [inline pool/tick.gno:19]: Is `consts.MAX_UINT128` immutable? if not, make it so.

**Review by dongwon8247 — CHANGES_REQUESTED** [review]
Add unit tests.

File: pool/tick.gno (hunk 1 above)
Comment (r3v4s) [reply pool/tick.gno]: 332de1e223fa9613049e9eb859b8649cd7c3004f

File: pool/tick.gno [hunk 2]
```diff
@@ -16,41 +16,23 @@ func tickTickSpacingToMaxLiquidityPerTick(tickSpacing int32) *u256.Uint {
...
 }
 
-// tickGetFeeGrowthInside calculates the fee growth inside a tick range.
+// tickGetFeeGrowthInside calculates the fee growth inside a tick range,
+// and returns the fee growth inside for both tokens.
 func (pool *Pool) tickGetFeeGrowthInside(
```
[the remark this answers is not in the document]
Comment (r3v4s) [reply pool/tick.gno]: 0b57caafc4ce4420062b715b2bd17d76f6a7c753

File: pool/tick.gno [hunk 3]
```diff
@@ -89,40 +70,75 @@ func (pool *Pool) tickUpdate(
...
-	info.feeGrowthOutside0X128 = new(u256.Uint).Sub(feeGrowthGlobal0X128, info.feeGrowthOutside0X128)
-	info.feeGrowthOutside1X128 = new(u256.Uint).Sub(feeGrowthGlobal1X128, info.feeGrowthOutside1X128)
+func (pool *Pool) getTick(tick int32) TickInfo {
+	tickInfo := pool.ticks[tick]
+	tickInfo.init()
+
+	return tickInfo
+}
+
+func getFeeGrowthBelowX128(
```
[the remark this answers is not in the document]
Comment (r3v4s) [reply pool/tick.gno]: 1f562d08ab4d39385a33dfa1bc916677aac44ae4

File: pool/tick.gno [hunk 4]
```diff
@@ -63,19 +45,18 @@ func (pool *Pool) tickUpdate(
...
 
-	info := pool.ticks[tick]
-	info.init()
+	thisTick := pool.getTick(tick)
 
-	liquidityGrossBefore := info.liquidityGross
+	liquidityGrossBefore := thisTick.liquidityGross
 	liquidityGrossAfter := liquidityMathAddDelta(liquidityGrossBefore, liquidityDelta)
 
 	if !(liquidityGrossAfter.Lte(maxLiquidity)) {
```
[the remark this answers is not in the document]
Comment (r3v4s) [reply pool/tick.gno:62]: panics reverts state, while err does update state and just returning error.

gno does embrace panic
- https://docs.gno.land/concepts/effective-gno/#embrace-panic

File: pool/tick.gno [hunk 5]
```diff
@@ -89,40 +70,65 @@ func (pool *Pool) tickUpdate(
...
+func (pool *Pool) getTick(tick int32) TickInfo {
+	tickInfo := pool.ticks[tick]
+	tickInfo.init()
+
+	return tickInfo
+}
 
-	info.feeGrowthOutside0X128 = new(u256.Uint).Sub(feeGrowthGlobal0X128, info.feeGrowthOutside0X128)
-	info.feeGrowthOutside1X128 = new(u256.Uint).Sub(feeGrowthGlobal1X128, info.feeGrowthOutside1X128)
+func getFeeGrowthX128(
```
Comment (onlyhyde) [inline pool/tick.gno:116]: I suggest this function should revert

This function is rather unreadable.
The traditional function seems more intuitive.
The conditional statement in this function tries to include the cases below and above, which makes it look more complicated with the separate conditional expressions.

**Review by onlyhyde — CHANGES_REQUESTED** [review]

File: pool/tick.gno (hunk 5 above)
Comment (r3v4s) [reply pool/tick.gno]: b69c6790916baf27d76fde2d936f99d0be5f7f20

File: pool/tick.gno [hunk 6]
```diff
@@ -9,48 +9,30 @@ import (
 	"gno.land/r/gnoswap/v1/consts"
 )
 
-// tickTickSpacingToMaxLiquidityPerTick calculates the maximum liquidity
+// calculateMaxLiquidityPerTick calculates the maximum liquidity
 // per tick for a given tick spacing.
-func tickTickSpacingToMaxLiquidityPerTick(tickSpacing int32) *u256.Uint {
+func calculateMaxLiquidityPerTick(tickSpacing int32) *u256.Uint {
```
[the remark this answers is not in the document]
Comment (r3v4s) [reply pool/tick.gno:14]: be7432442c24480e082e309cdffcecd26a5857f5

**Review by onlyhyde — APPROVED** [review]

**Review by notJoon — APPROVED** [review]

**Review by dongwon8247 — APPROVED** [review]

**Pull request merged**